src/gluonts/torch/model/deepar/_module.py

```diff
@@ -0,0 +1,323 @@
+        subsequences_length: int,
+    ) -> torch.Tensor:
+        """
+        Returns lagged subsequences of a given sequence.
+
+        Parameters
+        ----------
+        sequence : Tensor
+            the sequence from which lagged subsequences should be extracted.
+            Shape: (N, T, C).
+        subsequences_length : int
+            length of the subsequences to be extracted.
+
+        Returns
+        --------
+        lagged : Tensor
+            a tensor of shape (N, S, C, I), where S = subsequences_length and
+            I = len(indices), containing lagged subsequences. Specifically,
+            lagged[i, j, :, k] = sequence[i, -indices[k]-S+j, :].
+        """
+        sequence_length = sequence.shape[1]
+        indices = self.lags_seq
+
+        assert max(indices) + subsequences_length <= sequence_length, (
+            f"lags cannot go further than history length, found lag {max(indices)} "
+            f"while history length is only {sequence_length}"
+        )
+
+        lagged_values = []
+        for lag_index in indices:
+            begin_index = -lag_index - subsequences_length
+            end_index = -lag_index if lag_index > 0 else None
+            lagged_values.append(sequence[:, begin_index:end_index, ...])
+        return torch.stack(lagged_values, dim=-1)
+
+    def forward(
+        self,
+        prior_input: torch.Tensor,
+        input: torch.Tensor,
+        features: Optional[torch.Tensor] = None,
+        state: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+    ):
+        sequence = torch.cat((prior_input, input), dim=1)
+        lagged_target_sequence = self.get_lagged_subsequences(
+            sequence=sequence,
+            subsequences_length=input.shape[1],
+        )
+
+        expected_shape = input.shape[:2] + (
+            self.input_size * len(self.lags_seq),
+        )
+        assert (
+            lagged_target_sequence.shape == expected_shape
+        ), f"got shape {lagged_target_sequence.shape}, expected {expected_shape}"
+
+        if features is None:
+            rnn_input = lagged_target_sequence
+        else:
+            rnn_input = torch.cat((lagged_target_sequence, features), dim=-1)
+
+        if state is None:
+            return self.rnn(rnn_input)
+        else:
+            return self.rnn(rnn_input, state)
+
+
+class DeepARModel(nn.Module):
+    @validated()
+    def __init__(
+        self,
+        freq: str,
+        context_length: int,
+        prediction_length: int,
+        num_feat_dynamic_real: int,
+        num_feat_static_real: int,
+        num_feat_static_cat: int,
+        cardinality: List[int],
+        embedding_dimension: Optional[List[int]] = None,
+        num_layers: int = 2,
+        hidden_size: int = 40,
+        dropout_rate: float = 0.1,
+        distr_output: DistributionOutput = StudentTOutput(),
+        lags_seq: Optional[List[int]] = None,
+        scaling: bool = True,
+        num_parallel_samples: int = 100,
```

Comment:
I would not pass this as an initialization option here, but rather pass this to the `forward` method directly. The number of samples doesn't have an impact on the model architecture.

Reply:
True, on the other hand, the predictor that wraps only provides data to the model. A compromise is to leave it here, and allow an optional override in the `forward` signature